### PR TITLE
Reimplement placeholder <style> tag usage

### LINKED ISSUE #1845: Placeholder <style> tag ignored by rehydration logic
(I haven't provided a failing test PR because I'm unsure how to use Jest, apologies)

## Environment

`npx envinfo --system --binaries --npmPackages styled-components,babel-plugin-styled-components --markdown --clipboard` breaks, here's what it looks like:

```
npx: installed 1 in 1.783s
Path must be a string. Received undefined
npx: installed 1 in 1.849s
C:\Users\dylan\AppData\Roaming\npm-cache\_npx\4172\node_modules\envinfo\dist\cli.js

## System:
 - OS: Windows 10
 - CPU: x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
 - Memory: 6.35 GB / 15.89 GB
## Binaries:
 - Yarn: 1.6.0 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
 - npm: 5.6.0 - C:\Program Files\nodejs\npm.CMD

(node:4172) UnhandledPromiseRejectionWarning: Error: The system cannot find the path specified.

    at Function.e.exports.sync (C:\Users\[...]\AppData\Roaming\npm-cache\_npx\4172\node_modules\envinfo\dist\cli.js:2:7690)
    at Object.copySync (C:\Users\[...]\AppData\Roaming\npm-cache\_npx\4172\node_modules\envinfo\dist\cli.js:2:104342)
    at Object.t.writeSync.e [as writeSync] (C:\Users\[...]\AppData\Roaming\npm-cache\_npx\4172\node_modules\envinfo\dist\cli.js:2:122592)
    at C:\Users\[...]\AppData\Roaming\npm-cache\_npx\4172\node_modules\envinfo\dist\cli.js:2:123354
    at Promise.all.then.e (C:\Users\[...]\AppData\Roaming\npm-cache\_npx\4172\node_modules\envinfo\dist\cli.js:2:123369)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
(node:4172) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:4172) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

I'm using `styled-components@^3.3.3`

## Reproduction

https://codesandbox.io/s/oq1j5q52n9

Please inspect the head of the document in the iframe. It should look like this:
Rendered output

and the template document, you should be able to observe the following:
head snippet

## Steps to reproduce

1. Include styled-components in your react project (non-SSR), add a styled component    
2. Add the following tag to the head of the template React is putting `<App />` into: `<style data-styled-components="true"></style>` ([Documentation Source](https://www.styled-components.com/docs/advanced#issues-with-specificity))

## Expected Behavior

* styled-components styles should be placed in the placeholder css element
* this should be the same without SSR, as confirmed here: https://github.com/styled-components/styled-components/issues/1707#issuecomment-385006535

## Actual Behavior

* styled-components creates another css element at the bottom of `<head>`

### PR DESCRIPTION
**Pull request by cpmsmith**
This makes rehydration logic insert any new `<style>` tags immediately after the last original tag it found, effectively restoring the ability to use placeholder tags. V4 introduces one wrinkle, which is that the placeholder must have a matching `data-styled-version` attribute.

Fixes #1845.

### CODE REVIEW

File: packages/styled-components/src/models/StyleSheet.js [hunk 1]
```diff
@@ -111,13 +111,12 @@ export default class StyleSheet {
       els.push(el);
     }
 
-    /* abort rehydration if nothing was extracted */
     const extractedSize = extracted.length;
-    if (!extractedSize) return this;
-
```
Comment (kitten) [inline packages/styled-components/src/models/StyleSheet.js]: Yep this seems to be mostly the line at fault 😅

**Comment by probablyup** [comment]
I don't have a streaming setup anymore. @mxstbr think you could try this for spectrum?

**Comment by probablyup** [comment]
@cpmsmith could you add a changelog entry?

**Comment by cpmsmith** [comment]
@probablyup Yup, done

**Comment by probablyup** [comment]
@kitten do we still want this for v4?

**Comment by kitten** [comment]
@probablyup depends on whether we actually want another v4 release, I suppose?

**Comment by probablyup** [comment]
We're definitely doing at least 1 more. canary needs more baking time

**Comment by probablyup** [comment]
@kitten can you take another look at this? I'm not familiar with the code path being altered

**Comment by AlexMeah** [comment]
Any chance this could be merged? It's an incredibly useful when using SC with external stylesheets and libraries such as tailwind.

**Comment by brendonco** [comment]
This is a useful technique especially when you are working with Micro Frontend, where both host and standalone app have the same UI library with different version.

**Comment by kitten** [comment]
Closing due to age, inactivity, missing reproduction, or lack of details.

**Pull request closed without merging**